File: 0001-0500/0049. Group Anagrams.go
package main
// Medium
// Time:O(nk), Space:O(nk)
// https://leetcode.cn/problems/group-anagrams/

func groupAnagrams(strs []string) [][]string {
    groups := make(map[[26]int][]string)

    for _, s := range strs {
        // use char's freq as key
        key := [26]int{}
        for _, c := range s {
            key[c - 'a']++
        }
        groups[key] = append(groups[key], s)
    }
    res := [][]string{}
    for _, g := range groups {
        res = append(res, g)
    }
    return res
}

File: 0001-0500/0049. Group Anagrams.py
# Time:O(n*k), Space:O(n*k)

from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram = {}

        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            
            key = tuple(count)
            if key not in anagram:
                anagram[key] = []
            anagram[key].append(s)
        
        return list(anagram.values())
    